scripts/js-analyze.js

```diff
@@ -53,13 +53,36 @@ function locstr2(loc, str)
 
 function nameValid(name)
 {
-  return name &&
-         name.indexOf(" ") == -1 &&
-         name.indexOf("\n") == -1 &&
-         name.indexOf("\r") == -1 &&
-         name.indexOf("\0") == -1 &&
-         name.indexOf("\\") == -1 &&
-         name.indexOf('"') == -1;
+  if (!name) {
+    return false;
+  }
+  for (var i = 0; i < name.length; i++) {
+    var c = name.charCodeAt(i);
+    switch (c) {
+      case 0:  // '\0'
+      case 10: // '\n'
+      case 13: // '\r'
+      case 32: // ' '
+      case 34: // '"'
+      case 92: // '\\'
+        return false;
+    }
+    // Only allow valid surrogate pairs
+    if (c >= 0xD800 && c <= 0xDBFF && i + 1 < name.length) {
+      // c is a high surrogate, check to make sure next char is a low surrogate
+      var d = name.charCodeAt(i + 1);
+      if (d >= 0xDC00 && d <= 0xDFFF) {
+        // valid; skip over the pair and continue
+        i++;
+        continue;
+      }
+    }
+    // fail on any surrogate characters that weren't part of a pair
+    if (c >= 0xD800 && c <= 0xDFFF) {
```

Comment:
Looks like you could avoid having the `if (c >= 0xD800)` duplicated by early `continue`ing when `c` is less than 0xD800 and such... But maybe the code would be less obvious, so it's not a big deal.

Reply:
Ah good point, I made the change as it's probably worth doing the early-continue for the common case.